gnovm/cmd/gno/repl.go

```diff
@@ -107,68 +114,176 @@ func runRepl(cfg *replCfg) error {
 		handleInput(r, cfg.initialCommand)
 	}
 
-	fmt.Fprint(os.Stdout, "gno> ")
+	fmt.Fprint(os.Stdout, gnoREPL)
+
+	var (
+		inEdit      bool
+		prev        string
+		indentLevel int
+	)
 
-	inEdit := false
-	prev := ""
 	liner := bufio.NewScanner(os.Stdin)
 
 	for liner.Scan() {
 		line := liner.Text()
 
-		if l := strings.TrimSpace(line); l == ";" {
-			line, inEdit = "", false
-		} else if l == "/editor" {
-			line, inEdit = "", true
-			fmt.Fprintln(os.Stdout, "// enter a single ';' to quit and commit")
-		}
+		trimmedLine := strings.TrimSpace(line)
+
+		indentLevel = updateIndentLevel(trimmedLine, indentLevel)
+		line, inEdit = handleEditor(line)
+
 		if prev != "" {
 			line = prev + "\n" + line
 			prev = ""
 		}
+
 		if inEdit {
-			fmt.Fprint(os.Stdout, "...  ")
+			fmt.Fprint(os.Stdout, inEditMode)
 			prev = line
+
 			continue
 		}
 
 		if err := handleInput(r, line); err != nil {
 			var goScanError scanner.ErrorList
 			if errors.As(err, &goScanError) {
-				// We assune that a Go scanner error indicates an incomplete Go statement.
+				// We assume that a Go scanner error indicates an incomplete Go statement.
 				// Append next line and retry.
 				prev = line
 			} else {
 				fmt.Fprintln(os.Stderr, err)
 			}
 		}
 
-		if prev == "" {
-			fmt.Fprint(os.Stdout, "gno> ")
-		} else {
-			fmt.Fprint(os.Stdout, "...  ")
-		}
+		printPrompt(indentLevel, prev)
 	}
+
 	return nil
 }
 
+func handleEditor(line string) (string, bool) {
+	if l := strings.TrimSpace(line); l == ";" {
+		return "", false
+	} else if l == editorCommand {
+		fmt.Fprintln(os.Stdout, "// enter a single ';' to quit and commit")
+
+		return "", true
+	}
+
+	return line, false
+}
+
+func updateIndentLevel(line string, indentLevel int) int {
+	openCount, closeCount := 0, 0
+	increaseIndent := false
+
+	for i, char := range line {
```

Comment:
There are quite a few cases where this logic won't work as one would expect.
Example:
```go
gno> func main() {
...     v := "}}}}"
... | <--- cursor is back here again
```
Another example:
```go
gno> a := "{hello"

gno>     | <---- cursor moved up here
```

An easy way to "fix" this would be to just check the last characters of the line and apply the indentation changes. However, this won't work correctly if the user opens a block in a line followed by something on the same line that is in that block:
```go
gno> func DoSomething(){ i := 10
```
So maybe that isn't what we want, though it seems an odd thing to do.

The other solution is to also keep track of whether the rune being examined is quoted, either manually while iterating over runes or maybe a regular expression.

Reply:
I apologize for the delayed response. Firstly, I was aware of this issue, and it seems to have arisen because we simply counted the number of `{` or `}` characters.

As you suggested, using regex or adding rules or even creating a parser (though it might be overengineering) seemes necessary. I'll test this out and fix this issue with a good approach. thanks again!